panic: interface conversion: net.Conn is *tls.Conn, not *net.TCPConn

```
#./etcd -listen-client-urls=https://192.168.0.11:4002,http://127.0.0.1:4001 -listen-peer-urls=http://127.0.0.1:4003 -ca-file=/opt/my/ca.crt -cert-file=/opt/my/server.crt -key-file=/opt/my/server.key.insecure -advertise-client-urls=https://192.168.0.11:4002,http://127.0.0.1:4001 
2015/01/29 15:11:28 no data-dir provided, using default data-dir ./default.etcd
2015/01/29 15:11:28 etcd: listening for peers on http://127.0.0.1:4003
2015/01/29 15:11:28 etcd: clientTLS: cert = /opt/my/server.crt, key = /opt/my/server.key.insecure, ca = /opt/my/ca.crt
2015/01/29 15:11:28 etcd: listening for client requests on http://127.0.0.1:4001
2015/01/29 15:11:29 etcd: listening for client requests on https://192.168.0.11:4002
2015/01/29 15:11:29 etcdserver: name = default
2015/01/29 15:11:29 etcdserver: data dir = default.etcd
2015/01/29 15:11:29 etcdserver: heartbeat = 100ms
2015/01/29 15:11:29 etcdserver: election = 1000ms
2015/01/29 15:11:29 etcdserver: snapshot count = 10000
2015/01/29 15:11:29 etcdserver: advertise client URLs = http://127.0.0.1:4001,https://192.168.0.11:4002
2015/01/29 15:11:29 etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
2015/01/29 15:11:29 etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
2015/01/29 15:11:29 etcdserver: start member ce2a822cea30bfca in cluster 7e27652122e8b2ae
2015/01/29 15:11:29 raft: ce2a822cea30bfca became follower at term 0
2015/01/29 15:11:29 raft: newRaft ce2a822cea30bfca [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/01/29 15:11:29 raft: ce2a822cea30bfca became follower at term 1
panic: interface conversion: net.Conn is *tls.Conn, not *net.TCPConn

goroutine 12 [running]:
github.com/coreos/etcd/pkg/transport.(*keepaliveListener).Accept(0xc208031f90, 0x0, 0x0, 0x0, 0x0)
	/Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/pkg/transport/keepalive_listener.go:43 +0xe3
net/http.(*Server).Serve(0xc208062c00, 0x7fec19529510, 0xc208031f90, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1728 +0x92
github.com/coreos/etcd/etcdmain.serveHTTP(0x7fec19529510, 0xc208031f90, 0x7fec1952a7e0, 0xc20801cfc0, 0x0, 0x0, 0x0)
	/Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/http.go:36 +0x1be
github.com/coreos/etcd/etcdmain.func·006(0x7fec19529510, 0xc208031f90)
	/Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:186 +0x80
created by github.com/coreos/etcd/etcdmain.startEtcd
	/Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:187 +0x1e60
```

--- a/pkg/transport/keepalive_listener.go
+++ b/pkg/transport/keepalive_listener.go
@@ -15,6 +15,7 @@
 package transport
 
 import (
+	"crypto/tls"
 	"net"
 	"time"
 )
@@ -22,18 +23,26 @@ import (
 // NewKeepAliveListener returns a listener that listens on the given address.
 // http://tldp.org/HOWTO/TCP-Keepalive-HOWTO/overview.html
 func NewKeepAliveListener(addr string, scheme string, info TLSInfo) (net.Listener, error) {
-	ln, err := NewListener(addr, scheme, info)
+	l, err := net.Listen("tcp", addr)
 	if err != nil {
 		return nil, err
 	}
+
+	if !info.Empty() && scheme == "https" {
+		cfg, err := info.ServerConfig()
+		if err != nil {
+			return nil, err
+		}
+
+		return newTLSKeepaliveListener(l, cfg), nil
+	}
+
 	return &keepaliveListener{
-		Listener: ln,
+		Listener: l,
 	}, nil
 }
 
-type keepaliveListener struct {
-	net.Listener
-}
+type keepaliveListener struct{ net.Listener }
 
 func (kln *keepaliveListener) Accept() (net.Conn, error) {
 	c, err := kln.Listener.Accept()
@@ -47,4 +56,38 @@ func (kln *keepaliveListener) Accept() (net.Conn, error) {
 	tcpc.SetKeepAlive(true)
 	tcpc.SetKeepAlivePeriod(30 * time.Second)
 	return tcpc, nil
+}
+
+// A tlsKeepaliveListener implements a network listener (net.Listener) for TLS connections.
+type tlsKeepaliveListener struct {
+	net.Listener
+	config *tls.Config
+}
+
+// Accept waits for and returns the next incoming TLS connection.
+// The returned connection c is a *tls.Conn.
+func (l *tlsKeepaliveListener) Accept() (c net.Conn, err error) {
+	c, err = l.Listener.Accept()
+	if err != nil {
+		return
+	}
+	tcpc := c.(*net.TCPConn)
+	// detection time: tcp_keepalive_time + tcp_keepalive_probes + tcp_keepalive_intvl
+	// default on linux:  30 + 8 * 30
+	// default on osx:    30 + 8 * 75
+	tcpc.SetKeepAlive(true)
+	tcpc.SetKeepAlivePeriod(30 * time.Second)
+	c = tls.Server(c, l.config)
+	return
+}
+
+// NewListener creates a Listener which accepts connections from an inner
+// Listener and wraps each connection with Server.
+// The configuration config must be non-nil and must have
+// at least one certificate.
+func newTLSKeepaliveListener(inner net.Listener, config *tls.Config) net.Listener {
+	l := &tlsKeepaliveListener{}
+	l.Listener = inner
+	l.config = config
+	return l
 }

--- a/pkg/transport/keepalive_listener_test.go
+++ b/pkg/transport/keepalive_listener_test.go
@@ -15,7 +15,9 @@
 package transport
 
 import (
+	"crypto/tls"
 	"net/http"
+	"os"
 	"testing"
 )
 
@@ -34,4 +36,29 @@ func TestNewKeepAliveListener(t *testing.T) {
 		t.Fatalf("unexpected Accept error: %v", err)
 	}
 	conn.Close()
+	ln.Close()
+
+	// tls
+	tmp, err := createTempFile([]byte("XXX"))
+	if err != nil {
+		t.Fatalf("unable to create tmpfile: %v", err)
+	}
+	defer os.Remove(tmp)
+	tlsInfo := TLSInfo{CertFile: tmp, KeyFile: tmp}
+	tlsInfo.parseFunc = fakeCertificateParserFunc(tls.Certificate{}, nil)
+	tlsln, err := NewKeepAliveListener("127.0.0.1:0", "https", tlsInfo)
+	if err != nil {
+		t.Fatalf("unexpected NewKeepAliveListener error: %v", err)
+	}
+
+	go http.Get("https://" + tlsln.Addr().String())
+	conn, err = tlsln.Accept()
+	if err != nil {
+		t.Fatalf("unexpected Accept error: %v", err)
+	}
+	if _, ok := conn.(*tls.Conn); !ok {
+		t.Errorf("failed to accept *tls.Conn")
+	}
+	conn.Close()
+	tlsln.Close()
 }

